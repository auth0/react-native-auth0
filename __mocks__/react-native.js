/**
 * This file provides a project-wide, self-contained mock for the 'react-native' package.
 */

const React = require('react');

// A fake native module to be returned by TurboModuleRegistry.getEnforcing.
const mockNativeModule = {
  // ... (all the mock native methods from before)
  initializeAuth0WithConfiguration: jest.fn(() => Promise.resolve()),
  webAuth: jest.fn(() => Promise.resolve({})),
  webAuthLogout: jest.fn(() => Promise.resolve()),
  cancelWebAuth: jest.fn(() => Promise.resolve()),
  getBundleIdentifier: jest.fn(() => Promise.resolve('com.my.app')),
  resumeWebAuth: jest.fn(() => Promise.resolve(true)),
  saveCredentials: jest.fn(() => Promise.resolve()),
  getCredentials: jest.fn(() => Promise.resolve({})),
  hasValidCredentials: jest.fn(() => Promise.resolve(true)),
  clearCredentials: jest.fn(() => Promise.resolve()),
  hasValidInstance: jest.fn(() => Promise.resolve(true)),
};

// Export the mocked module.
module.exports = {
  // Mock the Platform module
  Platform: {
    OS: 'ios',
  },
  // Mock the TurboModuleRegistry
  TurboModuleRegistry: {
    getEnforcing: () => mockNativeModule,
  },
  // Mock the Linking module
  Linking: {
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
    getInitialURL: jest.fn(() => Promise.resolve(null)),
  },

  // FIX: Add mocks for core UI components used in tests.
  // We can use a simple functional component that just renders its children
  // and passes along any props.
  View: (props) => React.createElement('View', props, props.children),
  Text: (props) => React.createElement('Text', props, props.children),
  Button: (props) => React.createElement('Button', props, props.children),
  // Add any other components your tests might use, e.g., StyleSheet
  StyleSheet: {
    create: (styles) => styles,
  },
};
