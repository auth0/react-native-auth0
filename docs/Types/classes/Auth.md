[**react-native-auth0**](../../README.md)

---

[react-native-auth0](../../globals.md) / [Types](../README.md) / Auth

# Class: Auth

Defined in: [auth/index.ts:65](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L65)

Class for interfacing with the Auth0 Authentication API endpoints.

## See

https://auth0.com/docs/api/authentication

## Properties

### clientId

> `readonly` **clientId**: `string`

Defined in: [auth/index.ts:70](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L70)

The Auth0 client ID

---

### domain

> `readonly` **domain**: `string`

Defined in: [auth/index.ts:74](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L74)

The Auth0 tenant domain

## Methods

### authorizeUrl()

> **authorizeUrl**(`parameters`): `string`

Defined in: [auth/index.ts:104](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L104)

Builds the full authorize endpoint url in the Authorization Server (AS) with given parameters.

#### Parameters

##### parameters

[`AuthorizeUrlOptions`](../../interfaces/AuthorizeUrlOptions.md)

#### Returns

`string`

A URL to the authorize endpoint with specified parameters to redirect to for AuthZ/AuthN.

#### See

https://auth0.com/docs/api/authentication#authorize-client

---

### createUser()

> **createUser**(`parameters`): `Promise`\<`Partial`\<[`User`](../../type-aliases/User.md)\>\>

Defined in: [auth/index.ts:585](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L585)

Creates a new user using the options provided.

#### Parameters

##### parameters

[`CreateUserOptions`](../../interfaces/CreateUserOptions.md)

#### Returns

`Promise`\<`Partial`\<[`User`](../../type-aliases/User.md)\>\>

An instance of [User](../../type-aliases/User.md).

---

### exchange()

> **exchange**(`parameters`): `Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

Defined in: [auth/index.ts:149](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L149)

Exchanges a code obtained via `/authorize` (w/PKCE) for the user's tokens

#### Parameters

##### parameters

[`ExchangeOptions`](../../interfaces/ExchangeOptions.md)

#### Returns

`Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

A prominse for a populated instance of [Credentials](../../type-aliases/Credentials.md).

#### See

https://auth0.com/docs/api-auth/grant/authorization-code-pkce

---

### exchangeNativeSocial()

> **exchangeNativeSocial**(`parameters`): `Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

Defined in: [auth/index.ts:179](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L179)

Exchanges an external token obtained via a native social authentication solution for the user's tokens

#### Parameters

##### parameters

[`ExchangeNativeSocialOptions`](../../interfaces/ExchangeNativeSocialOptions.md)

#### Returns

`Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

A populated instance of [Credentials](../../type-aliases/Credentials.md).

#### See

https://auth0.com/docs/api/authentication#token-exchange-for-native-social

---

### loginWithEmail()

> **loginWithEmail**(`parameters`): `Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

Defined in: [auth/index.ts:305](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L305)

Completes the Passworldess authentication with an email connection that was started using [passwordlessWithEmail](#passwordlesswithemail).

#### Parameters

##### parameters

[`LoginWithEmailOptions`](../../interfaces/LoginWithEmailOptions.md)

#### Returns

`Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

A populated instance of [Credentials](../../type-aliases/Credentials.md).

---

### loginWithOOB()

> **loginWithOOB**(`parameters`): `Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

Defined in: [auth/index.ts:406](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L406)

Log in a user using an Out Of Band authentication code after they have received the 'mfa_required' error.
The MFA token tells the server the username or email, password, and realm values sent on the first request.

Requires your client to have the **MFA OOB** Grant Type enabled. See [Client Grant Types](https://auth0.com/docs/clients/client-grant-types) to learn how to enable it.

#### Parameters

##### parameters

[`LoginWithOOBOptions`](../../interfaces/LoginWithOOBOptions.md)

#### Returns

`Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

A populated instance of [Credentials](../../type-aliases/Credentials.md).

---

### loginWithOTP()

> **loginWithOTP**(`parameters`): `Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

Defined in: [auth/index.ts:373](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L373)

Log in a user using the One Time Password code after they have received the 'mfa_required' error.
The MFA token tells the server the username or email, password, and realm values sent on the first request.

Requires your client to have the **MFA OTP** Grant Type enabled.
See [Client Grant Types](https://auth0.com/docs/clients/client-grant-types) to learn how to enable it.

#### Parameters

##### parameters

[`LoginWithOTPOptions`](../../interfaces/LoginWithOTPOptions.md)

#### Returns

`Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

A populated instance of [Credentials](../../type-aliases/Credentials.md).

---

### loginWithRecoveryCode()

> **loginWithRecoveryCode**(`parameters`): `Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

Defined in: [auth/index.ts:440](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L440)

Log in a user using a multi-factor authentication Recovery Code after they have received the 'mfa_required' error.
The MFA token tells the server the username or email, password, and realm values sent on the first request.

Requires your client to have the **MFA** Grant Type enabled. See [Client Grant Types](https://auth0.com/docs/clients/client-grant-types) to learn how to enable it.

#### Parameters

##### parameters

[`LoginWithRecoveryCodeOptions`](../../interfaces/LoginWithRecoveryCodeOptions.md)

#### Returns

`Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

A populated instance of [Credentials](../../type-aliases/Credentials.md).

---

### loginWithSMS()

> **loginWithSMS**(`parameters`): `Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

Defined in: [auth/index.ts:337](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L337)

Completes the Passworldess authentication with an SMS connection that was started using [passwordlessWithSMS](#passwordlesswithsms).

#### Parameters

##### parameters

[`LoginWithSMSOptions`](../../interfaces/LoginWithSMSOptions.md)

#### Returns

`Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

A populated instance of [Credentials](../../type-aliases/Credentials.md).

---

### logoutUrl()

> **logoutUrl**(`parameters`): `string`

Defined in: [auth/index.ts:129](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L129)

Builds the full logout endpoint url in the Authorization Server (AS) with given parameters.

#### Parameters

##### parameters

[`LogoutUrlOptions`](../../interfaces/LogoutUrlOptions.md)

#### Returns

`string`

A URL to the logout endpoint with specified parameters

#### See

https://auth0.com/docs/api/authentication#logout

---

### multifactorChallenge()

> **multifactorChallenge**(`parameters`): `Promise`\<[`MultifactorChallengeResponse`](../../type-aliases/MultifactorChallengeResponse.md)\>

Defined in: [auth/index.ts:474](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L474)

Request a challenge for multi-factor authentication (MFA) based on the challenge types supported by the application and user.
The challenge type is how the user will get the challenge and prove possession. Supported challenge types include: "otp" and "oob".

#### Parameters

##### parameters

[`MultifactorChallengeOptions`](../../interfaces/MultifactorChallengeOptions.md)

#### Returns

`Promise`\<[`MultifactorChallengeResponse`](../../type-aliases/MultifactorChallengeResponse.md)\>

[MultifactorChallengeOTPResponse](../../type-aliases/MultifactorChallengeOTPResponse.md), [MultifactorChallengeOOBResponse](../../type-aliases/MultifactorChallengeOOBResponse.md), or [MultifactorChallengeOOBWithBindingResponse](../../type-aliases/MultifactorChallengeOOBWithBindingResponse.md) depending
on the challenge type.

---

### passwordlessWithEmail()

> **passwordlessWithEmail**(`parameters`): `Promise`\<`void`\>

Defined in: [auth/index.ts:262](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L262)

Starts the Passworldess flow with an email connection.

This should be completed later using a call to [loginWithEmail](#loginwithemail), passing the OTP that was sent to the user.

#### Parameters

##### parameters

[`PasswordlessWithEmailOptions`](../../interfaces/PasswordlessWithEmailOptions.md)

#### Returns

`Promise`\<`void`\>

---

### passwordlessWithSMS()

> **passwordlessWithSMS**(`parameters`): `Promise`\<`void`\>

Defined in: [auth/index.ts:279](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L279)

Starts the Passwordless flow with an SMS connection.

This should be completed later using a call to [loginWithSMS](#loginwithsms), passing the OTP that was sent to the user.

#### Parameters

##### parameters

[`PasswordlessWithSMSOptions`](../../interfaces/PasswordlessWithSMSOptions.md)

#### Returns

`Promise`\<`void`\>

---

### passwordRealm()

> **passwordRealm**(`parameters`): `Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

Defined in: [auth/index.ts:213](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L213)

Performs Auth with user credentials using the Password Realm Grant

#### Parameters

##### parameters

[`PasswordRealmOptions`](../../interfaces/PasswordRealmOptions.md)

#### Returns

`Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

A populated instance of [Credentials](../../type-aliases/Credentials.md).

#### See

https://auth0.com/docs/api-auth/grant/password#realm-support

---

### refreshToken()

> **refreshToken**(`parameters`): `Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

Defined in: [auth/index.ts:233](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L233)

Obtain new tokens using the Refresh Token obtained during Auth (requesting `offline_access` scope)

#### Parameters

##### parameters

[`RefreshTokenOptions`](../../interfaces/RefreshTokenOptions.md)

#### Returns

`Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

A populated instance of [Credentials](../../type-aliases/Credentials.md).

#### See

https://auth0.com/docs/tokens/refresh-token/current#use-a-refresh-token

---

### resetPassword()

> **resetPassword**(`parameters`): `Promise`\<`void`\>

Defined in: [auth/index.ts:566](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L566)

Request an email with instructions to change password of a user

#### Parameters

##### parameters

[`ResetPasswordOptions`](../../interfaces/ResetPasswordOptions.md)

#### Returns

`Promise`\<`void`\>

---

### revoke()

> **revoke**(`parameters`): `Promise`\<`void`\>

Defined in: [auth/index.ts:503](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L503)

Revoke an issued refresh token

#### Parameters

##### parameters

[`RevokeOptions`](../../interfaces/RevokeOptions.md)

#### Returns

`Promise`\<`void`\>

---

### userInfo()

> **userInfo**(`parameters`): `Promise`\<[`User`](../../type-aliases/User.md)\>

Defined in: [auth/index.ts:530](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/auth/index.ts#L530)

Return user information using an access token

#### Parameters

##### parameters

[`UserInfoOptions`](../../interfaces/UserInfoOptions.md)

#### Returns

`Promise`\<[`User`](../../type-aliases/User.md)\>

The user's profile information.
