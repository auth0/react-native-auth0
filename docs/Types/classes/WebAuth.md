[**react-native-auth0**](../../README.md)

---

[react-native-auth0](../../globals.md) / [Types](../README.md) / WebAuth

# Class: WebAuth

Defined in: [webauth/index.ts:24](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/webauth/index.ts#L24)

Helper to perform Auth against Auth0 hosted login page

It will use `/authorize` endpoint of the Authorization Server (AS)
with Code Grant and Proof Key for Challenge Exchange (PKCE).

## See

https://auth0.com/docs/api-auth/grant/authorization-code-pkce

## Methods

### authorize()

> **authorize**(`parameters`, `options`): `Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

Defined in: [webauth/index.ts:53](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/webauth/index.ts#L53)

Starts the AuthN/AuthZ transaction against the AS in the in-app browser.

To learn more about how to customize the authorize call, check the Universal Login Page
article at https://auth0.com/docs/hosted-pages/login

#### Parameters

##### parameters

[`WebAuthorizeParameters`](../../interfaces/WebAuthorizeParameters.md) = `{}`

##### options

[`WebAuthorizeOptions`](../../interfaces/WebAuthorizeOptions.md) = `{}`

#### Returns

`Promise`\<[`Credentials`](../../type-aliases/Credentials.md)\>

A poplulated instance of [Credentials](../../type-aliases/Credentials.md).

#### See

https://auth0.com/docs/api/authentication#authorize-client

---

### cancelWebAuth()

> **cancelWebAuth**(): `Promise`\<`void`\>

Defined in: [webauth/index.ts:79](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/webauth/index.ts#L79)

Terminates the ongoing web-based operation and reports back that it was cancelled.
You need to call this method within your custom Web Auth provider implementation whenever the operation is
cancelled by the user.

#### Returns

`Promise`\<`void`\>

---

### clearSession()

> **clearSession**(`parameters`, `options`): `Promise`\<`void`\>

Defined in: [webauth/index.ts:92](https://github.com/auth0/react-native-auth0/blob/64b3136e2ba68da80f979438fc7bc3abab9becdd/src/webauth/index.ts#L92)

Removes Auth0 session and optionally remove the Identity Provider session.

#### Parameters

##### parameters

[`ClearSessionParameters`](../../interfaces/ClearSessionParameters.md) = `{}`

##### options

[`ClearSessionOptions`](../../interfaces/ClearSessionOptions.md) = `{}`

#### Returns

`Promise`\<`void`\>

#### See

https://auth0.com/docs/logout
