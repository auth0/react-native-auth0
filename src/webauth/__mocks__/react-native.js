import Linking from './linking';
import A0Auth0 from './auth0';

const mock = {};

mock.Linking = new Linking();
mock.NativeModules = { A0Auth0: new A0Auth0() };
mock.Platform = { OS: 'test-os' };

module.exports = mock;
