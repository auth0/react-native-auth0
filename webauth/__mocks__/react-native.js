import Linking from './linking';
import A0Auth0 from './auth0';

const mock = {};

mock.Linking = new Linking();
mock.NativeModules = { A0Auth0: new A0Auth0() };
mock.Platform = { OS: "ios", Version: 11.1 };

module.exports = mock;