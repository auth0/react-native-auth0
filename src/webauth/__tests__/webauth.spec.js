jest.mock('react-native');
import Auth from '../../auth';
import WebAuth from '../index';
import {NativeModules} from 'react-native';
import {URL} from 'url';

const A0Auth0 = NativeModules.A0Auth0;

describe('WebAuth', () => {
  const auth = new Auth({baseUrl: 'https://auth0.com', clientId: 'abc123'});
  const webauth = new WebAuth(auth);

  beforeEach(() => {
    NativeModules.A0Auth0 = A0Auth0;
    A0Auth0.reset();
  });

  describe('clearSession', () => {
    it('should open log out URL', async () => {
      await webauth.clearSession();

      const parsedUrl = new URL(A0Auth0.url);
      expect(parsedUrl.protocol).toEqual('https:');
      expect(parsedUrl.hostname).toEqual('auth0.com');
      const urlQuery = parsedUrl.searchParams;
      expect(urlQuery.get('returnTo')).toEqual(
        'com.my.app://auth0.com/test-os/com.My.App/callback',
      );
      expect(urlQuery.get('client_id')).toEqual('abc123');
      expect(urlQuery.has('federated')).toEqual(false);
      expect(urlQuery.has('auth0Client')).toEqual(true);
    });

    it('should open log out URL with federated=true', async () => {
      const options = {federated: true};
      await webauth.clearSession(options);

      const parsedUrl = new URL(A0Auth0.url);
      expect(parsedUrl.protocol).toEqual('https:');
      expect(parsedUrl.hostname).toEqual('auth0.com');
      const urlQuery = parsedUrl.searchParams;
      expect(urlQuery.get('returnTo')).toEqual(
        'com.my.app://auth0.com/test-os/com.My.App/callback',
      );
      expect(urlQuery.get('client_id')).toEqual('abc123');
      expect(urlQuery.get('federated')).toEqual('true');
      expect(urlQuery.has('auth0Client')).toEqual(true);
    });
  });

  describe('organizations during log in', () => {
    it('should build the authorize URL with a organization', async () => {
      const newTransactionMock = jest
        .spyOn(webauth.agent, 'newTransaction')
        .mockImplementation(() =>
          Promise.resolve({state: 'state', verifier: 'verifier'}),
        );
      const showMock = jest
        .spyOn(webauth.agent, 'show')
        .mockImplementation(authorizeUrl => ({
          then: () => Promise.resolve(authorizeUrl),
        }));
      const parameters = {organization: 'test-org'};
      let url = await webauth.authorize(parameters, {});

      const parsedUrl = new URL(url);
      const urlQuery = parsedUrl.searchParams;
      expect(urlQuery.get('organization')).toEqual('test-org');
      newTransactionMock.mockRestore();
      showMock.mockRestore();
    });

    it('should build the authorize URL parsing organization from invitation URL', async () => {
      const newTransactionMock = jest
        .spyOn(webauth.agent, 'newTransaction')
        .mockImplementation(() =>
          Promise.resolve({state: 'state', verifier: 'verifier'}),
        );
      const showMock = jest
        .spyOn(webauth.agent, 'show')
        .mockImplementation(authorizeUrl => ({
          then: () => Promise.resolve(authorizeUrl),
        }));
      const parameters = {
        invitationUrl:
          'https://myapp.com/?invitation=inv123&organization=org123',
      };
      let url = await webauth.authorize(parameters, {});

      const parsedUrl = new URL(url);
      const urlQuery = parsedUrl.searchParams;
      expect(urlQuery.get('organization')).toEqual('org123');
      expect(urlQuery.get('invitation')).toEqual('inv123');
      newTransactionMock.mockRestore();
      showMock.mockRestore();
    });

    it('should throw an error when the values are not present in the invitation URL', async () => {
      const newTransactionMock = jest
        .spyOn(webauth.agent, 'newTransaction')
        .mockImplementation(() =>
          Promise.resolve({state: 'state', verifier: 'verifier'}),
        );
      const showMock = jest
        .spyOn(webauth.agent, 'show')
        .mockImplementation(authorizeUrl => ({
          then: () => Promise.resolve(authorizeUrl),
        }));

      expect.assertions(2);
      const parameters = {
        invitationUrl: 'https://myapp.com/?invitation=inv123',
      };
      await expect(webauth.authorize(parameters, {})).rejects.toMatchSnapshot();
      parameters.invitationUrl = 'https://myapp.com/?organization=org123';
      await expect(webauth.authorize(parameters, {})).rejects.toMatchSnapshot();

      newTransactionMock.mockRestore();
      showMock.mockRestore();
    });
  });

  describe('custom scheme', () => {
    it('should build the callback URL with a custom scheme when logging in', async () => {
      const newTransactionMock = jest
        .spyOn(webauth.agent, 'newTransaction')
        .mockImplementation(() =>
          Promise.resolve({state: 'state', verifier: 'verifier'}),
        );
      const showMock = jest
        .spyOn(webauth.agent, 'show')
        .mockImplementation(authorizeUrl => ({
          then: () => Promise.resolve(authorizeUrl),
        }));
      const options = {customScheme: 'custom-scheme'};
      let url = await webauth.authorize({}, options);

      const parsedUrl = new URL(url);
      const urlQuery = parsedUrl.searchParams;
      expect(urlQuery.get('redirect_uri')).toEqual(
        'custom-scheme://auth0.com/test-os/com.My.App/callback',
      );
      newTransactionMock.mockRestore();
      showMock.mockRestore();
    });

    it('should build the callback URL with a custom scheme when logging out', async () => {
      const options = {customScheme: 'custom-scheme'};
      await webauth.clearSession(options);

      const parsedUrl = new URL(A0Auth0.url);
      const urlQuery = parsedUrl.searchParams;
      expect(urlQuery.get('returnTo')).toEqual(
        'custom-scheme://auth0.com/test-os/com.My.App/callback',
      );
    });
  });
});
