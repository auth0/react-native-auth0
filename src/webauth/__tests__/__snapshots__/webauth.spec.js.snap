// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebAuth organizations during log in should throw an error when the values are not present in the invitation URL 1`] = `[a0.invalid_invitation_url: The invitation URL provided doesn't contain the 'organization' or 'invitation' values.]`;

exports[`WebAuth organizations during log in should throw an error when the values are not present in the invitation URL 2`] = `[a0.invalid_invitation_url: The invitation URL provided doesn't contain the 'organization' or 'invitation' values.]`;
